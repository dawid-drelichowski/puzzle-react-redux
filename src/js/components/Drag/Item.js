import React, {PropTypes} from 'react';

export default class extends React.Component {
  static propTypes = {
    field: PropTypes.number.isRequired
  };
  render() {
    const props = this.props;
    return <li draggable="true" data-field={props.field}/>;
  }
}
