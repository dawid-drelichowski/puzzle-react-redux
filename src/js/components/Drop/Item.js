import React, {PropTypes} from 'react';

export default class extends React.Component {
  static propTypes = {
    field: PropTypes.number.isRequired,
    currentlyDragged: PropTypes.number,
    onDrop: PropTypes.func.isRequired
  };
  static defaultProps = {
    currentlyDragged: 0
  };
  state = {dropped: false};
  onDragOver(event) {
    if (this.isValidDragElement()) {
      event.preventDefault();
    }
  }
  onDrop(event) {
    if (this.isValidDragElement()) {
      event.preventDefault();
      this.setState({dropped: true});
      this.props.onDrop(this.props.field);
    }
  }
  render() {
    let props = {};

    switch (this.state.dropped) {
      case true:
        props = {
          'data-field': this.props.field
        };
        break;
      default:
        props = {
          'data-valid': this.props.field,
          onDragOver: event => this.onDragOver(event),
          onDrop: event => this.onDrop(event)
        };
        break;
    }
    return <li {...props}></li>;
  }
  isValidDragElement() {
    return this.props.field === this.props.currentlyDragged;
  }
}
