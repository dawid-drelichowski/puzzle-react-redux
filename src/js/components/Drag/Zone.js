import React, {PropTypes} from 'react';
import Item from 'components/Drag/Item';

export default class extends React.Component {
  static propTypes = {
    piecesOrder: PropTypes.arrayOf(PropTypes.number).isRequired,
    onDragStart: PropTypes.func.isRequired,
    onDragEnd: PropTypes.func.isRequired
  };
  onDragStart = (event) => {
    const target = event.target,
      dataTransfer = event.dataTransfer;

    dataTransfer.effectAllowed = 'move';
    dataTransfer.dropEffect = 'move';
    dataTransfer.setData('text', '1'); //Firefox effectAllowed and dropEffect bug fix

    target.classList.add('draggable-current');
    this.props.onDragStart(this.getFieldFromElement(target));
  };
  onDragEnd = (event) => {
    const target = event.target;

    target.classList.remove('draggable-current');
    this.props.onDragEnd(this.getFieldFromElement(target));
  };
  render() {
    return <div className="col-lg-5 label label-danger">
      <ul
        className="grid label-danger"
        onDragStart={this.onDragStart}
        onDragEnd={this.onDragEnd}
      >
        {this.props.piecesOrder.map((value, index) => {
          return <Item key={index} field={value} />;
        })}
      </ul>
    </div>;
  }
  getFieldFromElement(element) {
    return parseInt(element.getAttribute('data-field'), 10);
  }
}
