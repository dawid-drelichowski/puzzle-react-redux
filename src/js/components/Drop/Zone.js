import React, {PropTypes} from 'react';
import Item from 'containers/Drop/Item';

export default class extends React.Component {
  static propTypes = {
    piecesCount: PropTypes.number.isRequired,
  };
  render() {
    const pieces = new Array(this.props.piecesCount).fill(true);

    return <div className="col-lg-5 label label-success">
      <ul className="grid label-success">
        {pieces.map((value, index) => {
          return <Item key={index} field={index + 1} />
        })}
      </ul>
    </div>;
  }
}