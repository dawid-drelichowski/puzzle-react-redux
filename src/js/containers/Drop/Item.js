import {connect} from 'react-redux';
import Creators from 'actions/index';
import Item from 'components/Drop/Item';

export default connect(
  state => {
    return {currentlyDragged: state.currentlyDragged};
  },
  dispatch => {
    return {
      onDrop: index => {
        dispatch(Creators.dropped(index));
      }
    };
  }
)(Item);
