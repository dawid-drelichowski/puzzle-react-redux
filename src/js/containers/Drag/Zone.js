import {connect} from 'react-redux';
import Creators from 'actions/index';
import Zone from 'components/Drag/Zone';

export default connect(
  () => {
    return {};
  },
  dispatch => {
    return {
      onDragStart: index => {
        dispatch(Creators.dragStart(index));
      },
      onDragEnd: index => {
        dispatch(Creators.dragEnd(index));
      }
    };
  }
)(Zone);
