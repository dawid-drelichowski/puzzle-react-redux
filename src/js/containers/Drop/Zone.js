import {connect} from 'react-redux';
import Creators from 'actions/index';
import Zone from 'components/Drop/Zone';

export default connect(
  state => {
    return {currentlyDragged: state.currentlyDragged};
  },
  dispatch => {
    return {
      onDrop: index => {
        dispatch(Creators.dropped(index));
      }
    };
  }
)(Zone);
