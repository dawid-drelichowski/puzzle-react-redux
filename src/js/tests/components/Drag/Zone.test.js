import React from 'react';
import Zone from 'components/Drag/Zone';
import Item from 'components/Drag/Item';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

describe('Drag zone component', () => {
  function getRenderedComponent(piecesOrder, onDragStart = () => {}, onDragEnd = () => {}) {
    return renderer.create(
      <Zone piecesOrder={piecesOrder} onDragStart={onDragStart} onDragEnd={onDragEnd} />
    );
  }

  function getMountedComponent(piecesOrder, onDragStart = () => {}, onDragEnd = () => {}) {
    return mount(
      <Zone piecesOrder={piecesOrder} onDragStart={onDragStart} onDragEnd={onDragEnd} />
    );
  }

  function getFakeDataTransfer() {
    return {dataTransfer: {setData: () => {}}};
  }

  it('should have 3 items', () => {
    expect(getRenderedComponent([3, 1, 2]).toJSON()).toMatchSnapshot();
  });

  it('should have 5 items', () => {
    expect(getRenderedComponent([3, 2, 1, 5, 4]).toJSON()).toMatchSnapshot();
  });

  it('item should have "draggable-current" class when drag started', () => {
    const item = getMountedComponent([3, 1, 2]).find(Item).at(1);

    item.simulate('dragStart', getFakeDataTransfer());
    expect(item.hasClass('draggable-current')).toBeTruthy();
  });

  it('item should not have "draggable-current" class when drag ended', () => {
    const item = getMountedComponent([3, 1, 2]).find(Item).at(0);

    item.simulate('dragEnd', getFakeDataTransfer());
    expect(item.hasClass('draggable-current')).toBeFalsy();
  });

  it('should call "onDragStart" callback when drag started', (done) => {
    const onDragStart = (index) => {
        expect(index).toBe(2);
        done();
      },
      item = getMountedComponent([3, 1, 2], onDragStart).find(Item).at(2);

    item.simulate('dragStart', getFakeDataTransfer());
  });

  it('should call "onDragEnd" callback when drag ended', (done) => {
    const onDragEnd = (index) => {
        expect(index).toBe(3);
        done();
      },
      item = getMountedComponent([3, 1, 2], () => {}, onDragEnd).find(Item).at(0);

    item.simulate('dragEnd', getFakeDataTransfer());
  });
});
