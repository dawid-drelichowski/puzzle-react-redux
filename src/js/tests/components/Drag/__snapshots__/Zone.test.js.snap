// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drag zone component should have 3 items 1`] = `
<div
  className="col-lg-5 label label-danger"
>
  <ul
    className="grid label-danger"
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <li
      data-field={3}
      draggable="true"
    />
    <li
      data-field={1}
      draggable="true"
    />
    <li
      data-field={2}
      draggable="true"
    />
  </ul>
</div>
`;

exports[`Drag zone component should have 5 items 1`] = `
<div
  className="col-lg-5 label label-danger"
>
  <ul
    className="grid label-danger"
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <li
      data-field={3}
      draggable="true"
    />
    <li
      data-field={2}
      draggable="true"
    />
    <li
      data-field={1}
      draggable="true"
    />
    <li
      data-field={5}
      draggable="true"
    />
    <li
      data-field={4}
      draggable="true"
    />
  </ul>
</div>
`;
